packages/opentelemetry-api/src/internal/global-utils.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { ContextManager } from '@opentelemetry/context-base';
+import type { TextMapPropagator } from '../context/propagation/TextMapPropagator';
+import type { MeterProvider } from '../metrics/MeterProvider';
+import { _globalThis } from '../platform';
+import type { TracerProvider } from '../trace/tracer_provider';
+import { VERSION } from '../version';
+import { isCompatible } from './semver';
+
+const GLOBAL_OPENTELEMETRY_API_KEY = Symbol.for('io.opentelemetry.js.api');
+
+const _global = _globalThis as OTelGlobal;
+
+export function registerGlobal<Type extends keyof OTelGlobalAPI>(
+  type: Type,
+  instance: OTelGlobalAPI[Type]
+): void {
+  _global[GLOBAL_OPENTELEMETRY_API_KEY] = _global[
+    GLOBAL_OPENTELEMETRY_API_KEY
+  ] ?? {
+    version: VERSION,
+  };
+
+  const api = _global[GLOBAL_OPENTELEMETRY_API_KEY]!;
+  if (api[type]) {
+    // already registered an API of this type
+    throw new Error(
```

Comment:
shouldn't we warn here only instead of throwing an error ?

Reply:
Yeah does the spec require us to never throw on the API ?

Reply:
The spec strongly discourages throwing. I can change it, but the behavior of doing nothing is potentially confusing. This should only be called during trace setup and never again so it seems unlikely to throw. I thought it would be fine in this case. I can change it to warn though if that's better.

Reply:
I would prefer a warn since users generally think there are errors anyway (since the console logger doesn't log the level by default)

Reply:
there is no logger here because we're in the API. It has to be `console.warn`. Is that OK?

Reply:
`console.error` ?

Reply:
That's fine too. The point is that it isn't using a logging configured by the user. Is that OK?

Reply:
I'm fine, however i would like to have some sort of prefix so end user know that the warning comes from otel.
A even better to do it would be to log a link to some kind of documentation that explains what to do in this case

Reply:
Logs on console may break command line tools

Reply:
@Flarna so does throwing :). Do you have a suggestion for when a logging interface isn't available? We could introduce a debug flag of some kind which enables these error logs similar to what the `debug` module does.

Reply:
> Logs on console may break command line tools

can you give some example, when in node it will break ?

Reply:
any CLI tool written in nodejs could potentially have an expected output to be parsed by other tools. If we log unexpectedly we change that output.

Simple example, a cli that scrapes the github API and outputs the open issues on a project as JSON to be consumed by `jq` and other tools.

Reply:
Exactly. It's not nodejs which runs into issues it's the specific user application.
I remember at least one case where I broke an application by logging to stderr because this was seen as error by the caller.
And logging to sdtout tends to break everything where a caller reads result from stdout.

If documentation clearly states that an API may throw it's up to the user (and all our samples) to add a try/catch and log/exit/...
A log to console can't be catched by user (except by monkey patching console but I don't think we want users to do this).

We could change the API to return a boolean instead of throwing. But this results in loosing the reason why it failed. We could return an error string or similar but I don't think these are nice APIs.

This is an API only used once during startup. Throwing there doesn't force users to clutter their whole codebase with try/catches. Therefore I think it's ok to throw here.

Reply:
> If documentation clearly states that an API may throw it's up to the user (and all our samples) to add a try/catch and log/exit/...

I don't think expecting users to try/catch when trying to setup the sdk or a instrumentation would be a good solution either.

Reply:
I see it as analogous to something like http server setup. If you try to bind to a port which is already bound, it throws. Because duplicate registration is startup/setup time it is most likely never throws or throws every time, not flaky or sometimes happening.

Reply:
Guys we can't break user application just because the otel configuration is wrong or because otel is missing something. Imagine the simple situation when you release a new version on production you then restart all the nodes and suddenly all production is down because the otel has wrong configuration. We can't do this. If otel fails all what should happen is to show user the error but never stop or break the application. The application should work fine in any situation.